File: backend/internal/models/tarea.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type Tarea struct {
	ID                   uuid.UUID  `json:"id" db:"id"`
	CursoID              uuid.UUID  `json:"curso_id" db:"curso_id"`
	TemaID               *uuid.UUID `json:"tema_id,omitempty" db:"tema_id"`
	Titulo               string     `json:"titulo" db:"titulo"`
	Descripcion          *string    `json:"descripcion,omitempty" db:"descripcion"`
	Semana               *int       `json:"semana,omitempty" db:"semana"`
	FechaPublicacion     time.Time  `json:"fecha_publicacion" db:"fecha_publicacion"`
	FechaLimite          time.Time  `json:"fecha_limite" db:"fecha_limite"`
	PuntajeMaximo        float64    `json:"puntaje_maximo" db:"puntaje_maximo"`
	PermiteEntregaTardia bool       `json:"permite_entrega_tardia" db:"permite_entrega_tardia"`
	PenalizacionPorDia   float64    `json:"penalizacion_por_dia" db:"penalizacion_por_dia"`
	DiasTolerancia       int        `json:"dias_tolerancia" db:"dias_tolerancia"`
	Tipo                 string     `json:"tipo" db:"tipo"`
	Activo               bool       `json:"activo" db:"activo"`
	CreatedAt            time.Time  `json:"created_at" db:"created_at"`

	// Stats para docente
	TotalEntregas        int `json:"total_entregas,omitempty"`
	EntregasSinCalificar int `json:"entregas_sin_calificar,omitempty"`
	EntregasCalificadas  int `json:"entregas_calificadas,omitempty"`
	EntregasPendientes   int `json:"entregas_pendientes,omitempty"`

	// Estado para estudiante
	MiEntrega      *Entrega `json:"mi_entrega,omitempty"`
	TiempoRestante *string  `json:"tiempo_restante,omitempty"`
	EstaVencida    bool     `json:"esta_vencida"`
}

type CreateTareaRequest struct {
	CursoID              uuid.UUID  `json:"curso_id" binding:"required"`
	TemaID               *uuid.UUID `json:"tema_id"`
	Titulo               string     `json:"titulo" binding:"required"`
	Descripcion          *string    `json:"descripcion"`
	Semana               *int       `json:"semana"`
	FechaLimite          time.Time  `json:"fecha_limite" binding:"required"`
	PuntajeMaximo        float64    `json:"puntaje_maximo" binding:"required,min=1"`
	PermiteEntregaTardia bool       `json:"permite_entrega_tardia"`
	PenalizacionPorDia   float64    `json:"penalizacion_por_dia"`
	DiasTolerancia       int        `json:"dias_tolerancia"`
	Tipo                 string     `json:"tipo" binding:"required,oneof=practica evaluacion proyecto"`
}

File: frontend/recetario/lib/data/models/tarea.dart
import 'entrega.dart';

class Tarea {
  final String id;
  final String cursoId;
  final String? temaId;
  final String titulo;
  final String? descripcion;
  final int? semana;
  final DateTime fechaPublicacion;
  final DateTime fechaLimite;
  final double puntajeMaximo;
  final bool permiteEntregaTardia;
  final double penalizacionPorDia;
  final int diasTolerancia;
  final String tipo; // practica, evaluacion, proyecto
  final bool activo;
  final DateTime createdAt;
  
  // Stats para docente
  final int? totalEntregas;
  final int? entregasSinCalificar;
  final int? entregasCalificadas;
  final int? entregasPendientes;
  
  // Para estudiante
  final Entrega? miEntrega;
  final String? tiempoRestante;
  final bool? estaVencida;

  Tarea({
    required this.id,
    required this.cursoId,
    this.temaId,
    required this.titulo,
    this.descripcion,
    this.semana,
    required this.fechaPublicacion,
    required this.fechaLimite,
    required this.puntajeMaximo,
    required this.permiteEntregaTardia,
    required this.penalizacionPorDia,
    required this.diasTolerancia,
    required this.tipo,
    required this.activo,
    required this.createdAt,
    this.totalEntregas,
    this.entregasSinCalificar,
    this.entregasCalificadas,
    this.entregasPendientes,
    this.miEntrega,
    this.tiempoRestante,
    this.estaVencida,
  });

  factory Tarea.fromJson(Map<String, dynamic> json) {
    return Tarea(
      id: json['id'],
      cursoId: json['curso_id'],
      temaId: json['tema_id'],
      titulo: json['titulo'],
      descripcion: json['descripcion'],
      semana: json['semana'],
      fechaPublicacion: DateTime.parse(json['fecha_publicacion']),
      fechaLimite: DateTime.parse(json['fecha_limite']),
      puntajeMaximo: json['puntaje_maximo'].toDouble(),
      permiteEntregaTardia: json['permite_entrega_tardia'] ?? false,
      penalizacionPorDia: json['penalizacion_por_dia']?.toDouble() ?? 0,
      diasTolerancia: json['dias_tolerancia'] ?? 0,
      tipo: json['tipo'] ?? 'practica',
      activo: json['activo'] ?? true,
      createdAt: DateTime.parse(json['created_at']),
      totalEntregas: json['total_entregas'],
      entregasSinCalificar: json['entregas_sin_calificar'],
      entregasCalificadas: json['entregas_calificadas'],
      entregasPendientes: json['entregas_pendientes'],
      miEntrega: json['mi_entrega'] != null
          ? Entrega.fromJson(json['mi_entrega'])
          : null,
      tiempoRestante: json['tiempo_restante'],
      estaVencida: json['esta_vencida'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'curso_id': cursoId,
      'tema_id': temaId,
      'titulo': titulo,
      'descripcion': descripcion,
      'semana': semana,
      'fecha_publicacion': fechaPublicacion.toUtc().toIso8601String(), // ✅ UTC
      'fecha_limite': fechaLimite.toUtc().toIso8601String(), // ✅ UTC
      'puntaje_maximo': puntajeMaximo,
      'permite_entrega_tardia': permiteEntregaTardia,
      'penalizacion_por_dia': penalizacionPorDia,
      'dias_tolerancia': diasTolerancia,
      'tipo': tipo,
    };
  }

  bool get yaEntregue => miEntrega != null;
  bool get estaCalificada => miEntrega?.calificacion != null;
  bool get estaPendiente => yaEntregue && !estaCalificada;
  
  String get estadoTexto {
    if (!yaEntregue) {
      if (DateTime.now().isAfter(fechaLimite)) return 'Vencida';
      return 'Pendiente de entrega';
    }
    if (estaCalificada) return 'Calificada';
    return 'Pendiente de calificación';
  }

  Duration? get tiempoHastaVencimiento {
    if (DateTime.now().isAfter(fechaLimite)) return null;
    return fechaLimite.difference(DateTime.now());
  }

  String get tiempoHastaVencimientoTexto {
    final tiempo = tiempoHastaVencimiento;
    if (tiempo == null) return 'Vencida';
    
    if (tiempo.inDays > 0) return 'Quedan ${tiempo.inDays} días';
    if (tiempo.inHours > 0) return 'Quedan ${tiempo.inHours} horas';
    return 'Quedan ${tiempo.inMinutes} minutos';
  }
}